fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/codec/LuceneOptimizedSegmentInfoFormat.java

```diff
@@ -0,0 +1,62 @@
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene.codec;
+
+import org.apache.lucene.codecs.SegmentInfoFormat;
+import org.apache.lucene.codecs.lucene70.Lucene70SegmentInfoFormat;
+import org.apache.lucene.index.IndexFileNames;
+import org.apache.lucene.index.SegmentInfo;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.IOContext;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.stream.Collectors;
+
+public class LuceneOptimizedSegmentInfoFormat extends SegmentInfoFormat {
+    private final SegmentInfoFormat segmentInfosFormat = new Lucene70SegmentInfoFormat();
+
+    public LuceneOptimizedSegmentInfoFormat() {
+    }
+
+    @Override
+    public SegmentInfo read(final Directory directory, final String segmentName, final byte[] segmentID, final IOContext context) throws IOException {
+        SegmentInfo segmentInfo = segmentInfosFormat.read(new LuceneOptimizedWrappedDirectory(directory), segmentName, segmentID, context);
+        SegmentInfo segmentInfoToReturn = new SegmentInfo(directory, segmentInfo.getVersion(), segmentInfo.getMinVersion(), segmentInfo.name,
+                segmentInfo.maxDoc(), segmentInfo.getUseCompoundFile(), segmentInfo.getCodec(), segmentInfo.getDiagnostics(),
+                segmentInfo.getId(),
+                segmentInfo.getAttributes(), segmentInfo.getIndexSort());
+        segmentInfoToReturn.setFiles(segmentInfo.files());
+        return segmentInfoToReturn;
+    }
+
+    @Override
+    public void write(final Directory dir, final SegmentInfo info, final IOContext ioContext) throws IOException {
+        SegmentInfo segmentInfoToWrite = new SegmentInfo(new LuceneOptimizedWrappedDirectory(dir), info.getVersion(), info.getMinVersion(), info.name,
+                info.maxDoc(), info.getUseCompoundFile(), info.getCodec(), Collections.emptyMap(),
+                info.getId(),
+                info.getAttributes(), info.getIndexSort());
+        segmentInfoToWrite.setFiles(info.files());
+        segmentInfosFormat.write(new LuceneOptimizedWrappedDirectory(dir), segmentInfoToWrite, ioContext);
+        final String fileName = IndexFileNames.segmentFileName(info.name, "", Lucene70SegmentInfoFormat.SI_EXTENSION);
+        info.setFiles(info.files().stream().filter(file -> !file.equals(fileName)).collect(Collectors.toSet()));
```

Comment:
Not to request change for this. Just wondering is this line just used to remove the ```si``` from existing segmentInfo in disk before this optimization?

Reply:
This is to prevent the format from saving the segment info because we are instead storing it within the file reference as bytes[].  the Info keeps track of all the files that are written to the segment and is used later at some places to read those files or validate their existence.

Reply:
Another related but issue (not with this pr) is that the segmentInfo actually marks the files (non compound) as deleted but they are still on the server when we query files/file sizes. It would be nice to clean these up. I think this happens because of lucene tombstones instead of hard delete calls.

Reply:
You are not seeing deletes of the initial non-compound files?  I would file a radar for someone to fix.